### DISCUSSION TITLE
MethodStatic keyword for local functions

### DISCUSSION THREAD

**Discussion by JKamsker** [dotnet/csharplang#3272] [Language Ideas] [answered]
The local function feature was introduced with C# 7.0. This allowed us to declare a method inside the body of another method.

I often come into situations where i don't want to access the context of the containing method, but the context of the containing class. 
Having two or more variables with the same name in function and local function could become dangerous when the declaration is deleted. In this case the local function would take the variable declared in the containing method.

Therefore a keyword which prohibits the local function accessing the context of the containing function, but allows it to access other members of the containing class would be useful.

Here an example:

```csharp
class MethodStaticExample
{
    private string _field;

    public MethodStaticExample()
    {
        _field = "Hello world";
    }

    public string DoSomeStuff(string input)
    {
        return DoFurtherStuff(input);

        //mstatic does not allow this local function to access `input` of the containing method, but accessing `_field` is ok
        mstatic string DoFurtherStuff(string input)
        {
            return string.Join(" ", $"{input} {_field}".Split(' ').Reverse());
        }
    }
 }
```

**Comment by gfoidl** [comment]
> but the context of the containing class.

Maybe that's an overuse of local functions --> just use a (regular) method instead?

**Comment by CyrusNajmabadi** [comment]
Or just pass 'this' into the local function.

**Comment by YairHalberstadt** [comment, marked as answer]
Now that attributes are allowed on local functions, this is a perfect scenario for an analyzer
